client/rpc-spec-v2/src/chain_head/chain_head_follow.rs

```diff
@@ -373,31 +373,26 @@ where
 				continue
 			}
 
-			match self.best_block_cache {
-				Some(best_block_hash) => {
-					// If the best reported block is a children of the last finalized,
-					// then we had a gap in notification.
-					let ancestor = sp_blockchain::lowest_common_ancestor(
-						&*self.client,
-						*last_finalized,
-						best_block_hash,
-					)?;
-
-					// A descendent of the finalized block was already reported
-					// before the `NewBlock` event containing the finalized block
-					// is reported.
-					if ancestor.hash == *last_finalized {
-						return Err(SubscriptionManagementError::Custom(
-							"A descendent of the finalized block was already reported".into(),
-						))
-					}
-					self.best_block_cache = Some(*hash);
-				},
-				// This is the first best block event that we generate.
-				None => {
-					self.best_block_cache = Some(*hash);
-				},
-			};
+			// Make sure we have not encountered a gap in notifications.
+			// Note: Let `generate_import_events()` populate the cache.
+			if let Some(best_block_hash) = self.best_block_cache {
+				// If the best reported block is a children of the last finalized,
+				// then we had a gap in notification.
+				let ancestor = sp_blockchain::lowest_common_ancestor(
+					&*self.client,
+					*last_finalized,
+					best_block_hash,
+				)?;
+
+				// A descendent of the finalized block was already reported
+				// before the `NewBlock` event containing the finalized block
+				// is reported.
+				if ancestor.hash == *last_finalized {
+					return Err(SubscriptionManagementError::Custom(
+						"A descendent of the finalized block was already reported".into(),
+					))
+				}
+			}
```

Comment:
I don't get this piece of code? How do you know that there wasn't any new best block event for the last finalized block?

And why do you have this as part of of the for loop while it doesn't depend on any from the loop?

Reply:
Hey

> How do you know that there wasn't any new best block event for the last finalized block?

We depend on the `pin_block` which returns `true` if this is the first time we "pin" the block and `false` if the block was previously pinned. We call `pin_block` twice for each block: from the import and finalized branches.

https://github.com/paritytech/substrate/blob/df6a43822dd150418832279ea1687d508d0f7e93/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs#L364-L369


> And why do you have this as part of of the for loop while it doesn't depend on any from the loop?

For any blocks that we have not seen yet we want to generate the NewBlock, except for the last block for which we want to generate the `NewBlock` + `BestBlock` events. For the last block, I've added the extra check to make sure we catch the following case:

```bash
[Block 0] -  [Block 1] - [Block 2] -  ...
                          ^^^ T0: Block import branch reports block 2 for the first time 

^^^^^^^^^^^^^^^ T1:  Finalized branch reports 0 and 1 for the first time
```

We could enter this path only if the BlockImport notification is not generated for 0, 1; but somehow we receive the  BlockImport for block 2. And at the same time we get Finalized for 0 and 1 (blocks that we have not been reported by BlockImport). 

This check may be an overcautious one, let me know what you think of it, thanks!